dev/io.openliberty.ejbcontainer.mdb.checkpoint_fat/fat/src/io/openliberty/ejbcontainer/mdb/ra/fat/tests/ConfigChangesTest.java

```diff
@@ -0,0 +1,285 @@
+
+        ShrinkHelper.exportAppToServer(server, MsgEndpointApp, DeployOptions.SERVER_ONLY);
+
+        server.addInstalledAppForValidation("MsgEndpointApp");
+
+        //#################### AdapterForEJB.jar (RAR implementation)
+        JavaArchive AdapterForEJBJar = ShrinkHelper.buildJavaArchive("AdapterForEJB.jar", "com.ibm.ws.ejbcontainer.fat.rar.*");
+        ShrinkHelper.exportToServer(server, "ralib", AdapterForEJBJar, DeployOptions.SERVER_ONLY);
+
+        //#################### AdapterForEJB.rar
+        ResourceAdapterArchive AdapterForEJBRar = ShrinkWrap.create(ResourceAdapterArchive.class, "AdapterForEJB.rar");
+        ShrinkHelper.addDirectory(AdapterForEJBRar, "test-resourceadapters/AdapterForEJB.rar/resources");
+        ShrinkHelper.exportToServer(server, "connectors", AdapterForEJBRar, DeployOptions.SERVER_ONLY);
+    }
+
+    @After
+    public void teardownTest() throws Exception {
+        switch (testMethod) {
+            case testMEPsActivateOnlyDuringRestoreBAS:
+            case testMEPsActivateOnlyDuringRestoreAAS:
+            case testAuthDataUpdatesDuringRestoreAAS:
+            case testJMSAuthDataUpdatesDuringRestoreAAS:
+            default:
+                if (server.isStarted()) {
+                    server.stopServer(IGNORE_REGEX);
+                }
+                break;
+        }
+    }
+
+    void runTest(String servlet) throws Exception {
+        FATServletClient.runTest(server, servlet, getTestMethodSimpleName());
+    }
+
+    /**
+     * Verify message endpoints do not activate during checkpoint beforeAppStart.
+     *
+     * This test verifies some JCA internals behavior, which may change as completing
+     * RAR installation at restore is not ideal.
+     */
+    @Test
+    public void testMEPsActivateOnlyDuringRestoreBAS() throws Exception {
+
+        // Verify RAR installation started during checkpoint BAS
+        testMsgs = server.findStringsInLogsUsingMark("J2CA7018I: .*AdapterForEJB",
+                                                     server.getDefaultLogFile());
+        assertFalse(testMsgs.isEmpty());
+
+        // Checkpoint BAS executes when the JCA runtime loads RA implementation classes
+        // while processing RAR metadata, i.e., RAR annotations and java bean properties.
+
+        // Verify RAR installation did not complete
+        testMsgs = server.findStringsInLogsUsingMark("J2CA7001I: .*AdapterForEJB",
+                                                     server.getDefaultLogFile());
+        assertTrue(testMsgs.isEmpty());
+
+        // Verify checkpoint did not execute while starting (loading) applications,
+        // further ensuring no endpoint has initialized nor activated
+        testMsgs = server.findStringsInLogsUsingMark("CWWKZ0018I: .*MsgEndpointApp",
+                                                     server.getDefaultLogFile());
+        assertTrue(testMsgs.isEmpty());
+
+        server.checkpointRestore();
+    }
```

Comment:
I see negative assertions before restore to make sure some things do not happen before restore when using `beforeAppStart`, but should there be positive assertions after restore that make sure these things do happen after restore?

Do we need an `afterAppStart` test that makes sure these things happen before checkpoint completes?  Or do these things always complete after restore regardless of the checkpoint phase?  If so it seem there should still be a test that makes sure that happens correctly for `afterAppStart

Reply:
Yes, this test and `testMEPsActivateOnlyDuringRestoreAAS` will benefit from explicit assertions that endpoints activate during `server.checkpointRestore()` and only after server startup.  These tests were otherwise relying on the implicit assertion that no unexpected failure occurs and that other tests verify endpoints activate. Since no test in the suite verifies endpoints activate in restore of checkpoint-at-BAS, I'll add the assertions to `testMEPsActivateOnlyDuringRestoreBAS` and to `testMEPsActivateOnlyDuringRestoreAAS`.

Regarding your 2nd question, I don't think we need another test.  Adding assertions to clearly verify endpoints activate only after server startup will suffice.  Some of the assertions in the BAS test verify JCA internals that will be impacted if we modify checkpoint classloading injection for resource binaries, and the impact is specific to checkpoint at BAS.